**Απόστολος Μουστάκης**

**AM: 03400182**

<font color='darkred'>Οι απαντήσεις σε όλα τα ερωτήματα γράφονται με αυτό το χρώμα ώστε να ξεχωρίζουν από τις εκφωνήσεις </font>

<h1>Q-Learning</h1>

<p>Στην συγκεκριμένη άσκηση θα μελετήσετε το στοχαστικό αλγόριθμο Q-Learning, χρησιμοποιώντας το έτοιμο πρόγραμμα που σας δίνεται.</p> <p> Στα πλαίσια του παραδείγματος θα εξετάσετε μία υλοποίηση με Q-learning σχετικά με το σύστημα αυτόματης οδήγησης ενός ταξί. Στα πλαίσια του προβλήματος αυτού θα πρέπει να ικανοποιούνται τα εξής:</p>
<ul>
<li>Το ταξί θα πρέπει να αφήνει τον πελάτη στη σωστή θέση</li>
<li>Το ταξί να ακολουθεί τη συντομότερη δυνατή διαδρομή</li>
<li>Να τηρούνται οι κανόνες κυκλοφορίας και ασφάλειας των επιβατών</li>
</ul>

<p>Στα πλαίσια του προβλήματος έχουμε τα εξής χαρακτηριστικά για την επιβράβευση, τις καταστάσεις και τις ενέργειες.</p> 

<h4>Επιβράβευση</h4>
<ul>
<li>Θα έχουμε τη μέγιστη επιβράβευση όταν το ταξί αφήνει έναν πελάτη στην επιθυμητή θέση</li>
<li>Θα υπάρχει ποινή στην περίπτωση όπου το ταξί αφήσει τον πελάτη σε κάποιο λανθασμένο σημείο</li>
<li>Ο agent θα παίρνει μία μικρή σχετικά ποινή στην περίπτωση όπου αργεί να φτάσει στον τελικό προορισμό</li>
</ul>

<p>Γενικά οι παραπάνω αρχές συνοψίζονται στα εξής: "Λαμβάνουμε +20 πόντους για μια επιτυχημένη πτώση και χάνουμε 1 πόντο για κάθε χρονικό βήμα που παίρνει. Υπάρχει επίσης ποινή 10 πόντων για παράνομες ενέργειες παραλαβής και αποχώρησης."</p>

<h4>Πλήθος Καταστάσεων</h4>
<img src="https://storage.googleapis.com/lds-media/images/Reinforcement_Learning_Taxi_Env.width-1200.png">
<p>Στο παράδειγμά μας έχουμε ένα μικρό χώρο 5*5. Από εκεί και πέρα έχουμε 4 προορισμούς και 5 πιθανές θέσεις του πελάτη (παίρνουμε και την περίπτωση ο πελάτης να είναι ήδη μέσα στο τάξι).</p>
<p>Με βάση τα παραπάνω έχουμε 5*5*5*4=500 πιθανές καταστάσεις.</p>

<h4>Πλήθος Ενεργειών</h4>
<p>Έχουμε έξι ενεργειες για το ταξί, οι οποίες είναι οι εξής:</p>
<ul>
<li>0=Νότια</li>
<li>1=Βόρεια</li>
<li>2=Ανατολικά</li>
<li>3=Δυτικά</li>
<li>4=Επιβίβαση</li>
<li>5=Αποβίβαση</li>
</ul>





<p>Πριν συνεχίσετε στην άσκηση να απαντήσετε στο εξής ερώτημα:</p>

<b><p>1. Να περιγράψετε σύντομα τον αλγόριθμο Q-Learning. Σε ποια προβλήματα θεωρείτε ότι ταιριάζει ως τρόπος εκμάθησης η Ενισχυτική Μάθηση (Reinforcement Learning); Ποια είναι η βασική διαφορά του αλγορίθμου Q-Learning από τους αλγορίθμους Policy Iteration και Value Iteration;</p></b>


<font color='darkred'>Η βασική ιδέα του Q-learning είναι ότι υπολογίζουμε την μελλοντική αναμενόμενη επιβράβευση για κάθε δράση σε κάθε κατάσταση. Αυτό ονομάζεται Q-value (ένα ζεύγος κατάσταση-δράση). Με αυτόν τον τρόπο μπορούμε γρήγορα και εύκολα να ανακτήσουμε μία πολιτική: σε μία κατάσταση s απλά επιλέγουμε την δράση με το μεγαλύτερο Q-value.
$$π'(s) = arg\max_a Q(s,a)$$
Ο αλγόριθμος Q-learning μοντελοποιείται μαθηματικά ως εξής:
1.   Αρχικοποίησε $Q(s,a)=0$ για όλα τα ζευγάρια κατάσταση-δράση
2.   Στην κατάσταση $s$ κάνε την δράση $α=π’(s)$, δηλαδή την καλύτερη δράση, ώστε να καταλήξεις σε κατάσταση $s’$ με επιβράβευση $r$ 
3.	Υπολόγισε την νέα εκτιμώμενη Q-value:
$$Q'(s,a) = r + γ \max_{α'} Q_k(s',a') $$
2.   Ενημερώστε την τρέχουσα Q-value εξομαλύνοντας τη νέα εκτιμώμενη και την υπάρχουσα:
$$Q_{k+1}(s,a) = (1-a)Q_k(s,a) + a Q'(s,a) $$
1.   Πήγαινε στο βήμα 2 μέχρι να επέλθει σύγκλιση

<font color='darkred'>Η Ενισχυτική Μάθηση (Reinforcement Learning) ταιριάζει ως τρόπος εκμάθησης σε προβλήματα όπου έχουν σαν στόχο να εκπαιδευτεί ένας αλγόριθμος που θα προσαρμόζει την συμπεριφορά του σε μία συγκεκριμένη κατάσταση. Ο αλγόριθμος αυτός εξερευνάει το περιβάλλον παίρνοντας αποφάσεις και αποκτώντας εμπειρία μέσω της ανατροφοδότησης του με αυτό, η οποία δίνεται με την μορφή επιβραβεύσεων και κυρώσεων. Χαρακτηριστικό παράδειγμα εφαρμογής της Ενισχυτικής Μάθησης είναι σε παιχνίδια, όπως το σκάκι και το Go, όπου ο αλγόριθμος κερδίζει τους καλύτερους παίκτες στον κόσμο.

<font color='darkred'>Η κύρια διαφορά του Q-Learning από τους αλγορίθμους Policy και Value Iteration είναι ότι ο πράκτορας δεν γνωρίζει τις πιθανότητες μετάβασης των καταστάσεων και τις ανταμοιβές. Ο πράκτορας ανακαλύπτει ότι υπάρχει μία ανταμοιβή για την μετάβαση από μία κατάσταση σε μία άλλη μέσω μίας δεδομένης δράσης μόνο όταν κάνει την συγκεκριμένη δράση και λάβει την ανταμοιβή. Ομοίως, υπολογίζει μόνο ποιες μεταβάσεις είναι διαθέσιμες από μία δεδομένη κατάσταση όταν καταλήγει σε αυτή την κατάσταση και εξετάζει τις επιλογές του. Ένα οι καταστάσεις είναι στοχαστικές μαθαίνει την πιθανότητα μετάβασης μεταξύ των καταστάσεων παρατηρώντας πόσο συχνά συμβαίνουν διαφορετικές μεταβάσεις. Ο αλγόριθμος Q-Learning είναι model free: Δεν πραγματοποιεί υποθέσεις για το περιβάλλον αλλά το εξερευνάει. 






<p>Στη συνέχεια θα πρέπει να φορτώσετε τη βιβλιοθήκη gym καθώς και το σχετικό dataset<p>

In [ ]:
!pip uninstall -y gym
!pip install setuptools==66
!pip install gym==0.18.0
!pip install gym[atari]
!pip install cmake scipy

In [ ]:
import gym

env = gym.make("Taxi-v3").env

env.render()

+---------+
|R: | : :G|
| : | : : |
| : : : : |
| | : | : |
|Y| : |B: |
+---------+



In [ ]:
env.reset() # reset environment to a new, random state
env.render()

print("Action Space {}".format(env.action_space))
print("State Space {}".format(env.observation_space))


+---------+
|R: | : :G|
| : | : : |
| : : : : |
| | : | : |
|Y| : |B: |
+---------+

Action Space Discrete(6)
State Space Discrete(500)


<p>Παρακάτω ορίζουμε τις συνεταγμένες του ταξί, τη θέση του πελάτη και το σημείο προορισμού</p>

In [ ]:
state = env.encode(3, 1, 2, 0) # (taxi row, taxi column, passenger index, destination index)
print("State:", state)

env.s = state
env.render()

State: 328
+---------+
|R: | : :G|
| : | : : |
| : : : : |
| | : | : |
|Y| : |B: |
+---------+



<p>Παρακάτω είναι η μήτρα επιβράβευσης για το state που ορίσαμε στο προηγούμενο βήμα</p>

In [ ]:
env.P[328]

{0: [(1.0, 428, -1, False)],
 1: [(1.0, 228, -1, False)],
 2: [(1.0, 348, -1, False)],
 3: [(1.0, 328, -1, False)],
 4: [(1.0, 328, -10, False)],
 5: [(1.0, 328, -10, False)]}

<p> Τρέχουμε το παράδειγμά μας χωρις τη χρήση Q-Learning.</p>

<b><p>2. Τα αποτελέσματα είναι ικανοποιητικά; Πως θα μας εξυπηρετούσε η χρήση του Q-Learning;</p></b>

<font color='darkred'>Προφανώς τα αποτελέσματα δεν είναι καθόλου ικανοποιητικά. Ο πράκτορας κάνει πάρα πολλά βήματα και ένα μεγάλο πλήθος από αυτά αντιστοιχεί σε σφάλματα (penalties). Συγκεκριμένα κάθε φορά περίπου το 1/3 των βημάτων αντιστοιχεί σε σφάλματα. Αυτό συμβαίνει επειδή ο πράκτορας δεν έχει την δυνατότητα να μάθει από την εμπειρία του παρελθόντος και να την αξιοποιήσει ώστε να παίρνει καλύτερες αποφάσεις. Όσες φορές και να τρέξουμε τον αλγόριθμο δεν θα βελτιστοποιηθεί ποτέ. Με την χρήση του Q-learning ο πράκτορας θα χρησιμοποιήσει τις ανταμοιβές του περιβάλλοντος για να μάθει με την πάροδο του χρόνου την καλύτερη ενέργεια που πρέπει να κάνει σε μία δεδομένη κατάσταση. Έτσι θα έχουμε πολύ καλύτερα αποτελέσματα.

In [ ]:
env.s = 328  # set environment to illustration's state

epochs = 0
penalties, reward = 0, 0

frames = [] # for animation

done = False

while not done:
    action = env.action_space.sample()
    state, reward, done, info = env.step(action)

    if reward == -10:
        penalties += 1
    
    # Put each rendered frame into dict for animation
    frames.append({
        'frame': env.render(mode='ansi'),
        'state': state,
        'action': action,
        'reward': reward
        }
    )

    epochs += 1
    
    
print("Timesteps taken: {}".format(epochs))
print("Penalties incurred: {}".format(penalties))

Timesteps taken: 960
Penalties incurred: 344


<p>Τώρα θα προσπαθήσουμε να λύσουμε το πρόβλημά μας με τη χρήση του Q-Learning.</p>

<b><p>3. Τι γνωρίζετε για τις παραμέτρους α και γ. Τι θα συμβεί αν έχουν τιμές ίσες με 1;</p></b>

<font color='darkred'>Η παράμετρος $γ$ αφορά τον παράγοντα έκπτωσης (discount factor) και καθορίζει πόση σημασία δίνεται σε μελλοντικές επιβραβεύσεις. Ισχύει $0 \leq γ \leq 1$. Μία υψηλή τιμή του $γ$, κοντά στο 1, δηλώνει πως δίνεται μεγάλη βαρύτητα σε μελλοντικές επιβραβεύσεις, ενώ μία χαμηλή τιμή του $γ$, κοντά στο 0, σημαίνει πως ο πράκτορας δίνει περισσότερη σημασία στις άμεσες επιβραβεύσεις και συνεπώς ο αλγόριθμος γίνεται πιο άπληστος. Στην ακραία περίπτωση όπου $γ = 1$ ο αλγόριθμος δεν συγκλίνει, καθώς η value function αποκλίνει.

<font color='darkred'>Η παράμετρος $α$ είναι ο ρυθμός εκμάθησης, ο οποίος αφορά τον βαθμό στον οποίο οι Q-values ενημερώνονται σε κάθε επανάληψη. Ισχύει $0 \leq α \leq 1$.  Στην περίπτωση όπου $α = 0$ ο πράκτορας δεν μαθαίνει καθόλου από το περιβάλλον, καθώς οι Q-values δεν ανανεώνονται. Από την άλλη, όσο το $α$ πλησιάζει το 1 ο αλγόριθμος μαθαίνει πολύ γρήγορα. Στην ακραία περίπτωση όπου $α = 1$ οι νέες Q-values θα παρακάμψουν εντελώς τις παλιές και συνεπώς ο αλγόριθμος δεν κρατάει μνήμη. Αυτό μπορεί να οδηγήσει σε Q-values οι οποίες είναι ασταθείς και ταλαντεύονται μεταξύ τιμών εμποδίζοντας τον αλγόριθμο στο να συγκλίνει σε βέλτιστη λύση.


In [ ]:
import numpy as np
q_table = np.zeros([env.observation_space.n, env.action_space.n])

In [ ]:
%%time
"""Training the agent"""

import random
from IPython.display import clear_output

# Hyperparameters
alpha = 0.1
gamma = 0.6
epsilon = 0.1

# For plotting metrics
all_epochs = []
all_penalties = []

for i in range(1, 100001):
    state = env.reset()

    epochs, penalties, reward, = 0, 0, 0
    done = False
    
    while not done:
        if random.uniform(0, 1) < epsilon:
            action = env.action_space.sample() # Explore action space
        else:
            action = np.argmax(q_table[state]) # Exploit learned values

        next_state, reward, done, info = env.step(action) 
        
        old_value = q_table[state, action]
        next_max = np.max(q_table[next_state])
        
        new_value = (1 - alpha) * old_value + alpha * (reward + gamma * next_max)
        q_table[state, action] = new_value

        if reward == -10:
            penalties += 1

        state = next_state
        epochs += 1
        
    if i % 100 == 0:
        clear_output(wait=True)
        print(f"Episode: {i}")

print("Training finished.\n")

Episode: 100000
Training finished.

CPU times: user 1min 2s, sys: 7.83 s, total: 1min 10s
Wall time: 1min 7s


In [ ]:
q_table[328]

array([ -2.41505176,  -2.27325184,  -2.40231418,  -2.35947929,
        -9.77058538, -10.20595373])

In [ ]:
"""Evaluate agent's performance after Q-learning"""

total_epochs, total_penalties = 0, 0
episodes = 100
rewards=[]

for _ in range(episodes):
    state = env.reset()
    epochs, penalties, reward = 0, 0, 0
    
    done = False
    
    while not done:
        action = np.argmax(q_table[state])
        state, reward, done, info = env.step(action)
        rewards.append(reward)

        if reward == -10:
            penalties += 1

        epochs += 1

    total_penalties += penalties
    total_epochs += epochs

rewards = np.asarray(rewards)
print(f"Results after {episodes} episodes:")
print(f"Average timesteps per episode: {total_epochs / episodes}")
print(f"Average penalties per episode: {total_penalties / episodes}")
print(f"Average reward per step: {rewards.mean():.2f}")

Results after 100 episodes:
Average timesteps per episode: 12.98
Average penalties per episode: 0.0
Average reward per step: 0.62


<b><p>4. Συγκρίνετε τους δύο αλγορίθμους με βάση τις παρακάτω μετρικές</p>
<ul>
<li>Μέσος αριθμός παραβάσεων ανά επεισόδιο</li>
<li>Μέσος αριθμός βημάτων ανά διαδρομή</li>
<li>Μέσος αριθμός ανταμοιβών ανά κίνηση</li>
</ul>
<p>Τις παραπάνω συγκρίσεις να τις κάνετε για 100 επεισόδια.</p>
</b>

In [ ]:
"""Evaluate agent's performance before Q-learning"""

total_epochs, total_penalties = 0, 0
episodes = 100
rewards=[]

for i in range(episodes):
  env.s = 328  # set environment to illustration's state

  epochs = 0
  penalties, reward = 0, 0

  frames = [] # for animation

  done = False

  while not done:
      action = env.action_space.sample()
      state, reward, done, info = env.step(action)
      rewards.append(reward)

      if reward == -10:
         penalties += 1
    
      #Put each rendered frame into dict for animation
      frames.append({
          'frame': env.render(mode='ansi'),
          'state': state,
          'action': action,
          'reward': reward
          }
      )

      epochs += 1

  total_penalties += penalties
  total_epochs += epochs

rewards = np.asarray(rewards)
print(f"Results after {episodes} episodes:")
print(f"Average timesteps per episode: {total_epochs / episodes}")
print(f"Average penalties per episode: {total_penalties / episodes}")
print(f"Average reward per step: {rewards.mean():.2f}")


Results after 100 episodes:
Average timesteps per episode: 1837.63
Average penalties per episode: 596.39
Average reward per step: -3.91


<font color='darkred'>Η σύγκριση των αλγορίθμων γίνεται με τις παρακάτω μετρικές:
*  Μέσος αριθμός παραβάσεων ανά επεισόδιο: Όσο μικρότερος είναι ο αριθμός αυτός τόσο καλύτερη είναι η απόδοση του agent. Ιδανικά θα θέλαμε η μέτρηση αυτή να είναι 0 ή πολύ κοντά στο 0.
*  Μέσος αριθμός βημάτων ανά διαδρομή: Επίσης θέλουμε αυτός ο αριθμός να είναι μικρός, καθώς θέλουμε ο agent να κάνει τα ελάχιστα δυνατά βήματα για να φτάσει στον προορισμό (shortest path).
*  Μέσος αριθμός ανταμοιβών ανά κίνηση: Όσο μεγαλύτερη είναι η ανταμοιβή σημαίνει ότι ο agent κάνει το σωστό. Στην συγκεκριμένη περίπτωση, όπου τα timesteps και τα penalties ανταμείβονται αρνητικά, μια υψηλότερη μέση ανταμοιβή θα σήμαινε ότι ο agent φτάνει στον προορισμό του όσον τον δυνατόν γρηγορότερα και με τις λιγότερες κυρώσεις. </font>

<font color='darkred'>Τα συγκεντρωτικά αποτελέσματα των δύο αλγορίθμων για 100 επεισόδια φαίνονται στον παρακάτω πίνακα. Είναι εμφανές ότι ο αλγόριθμος Q-Learning έχει καταπληκτικά αποτελέσματα.
<center>

||  Χωρίς Q-Learning (Random)| Με Q-Learning | 
| :----:       |    :----:   |   :----:   |     
| Μέσος αριθμός παραβάσεων ανά επεισόδιο|596.39|0.0
| Μέσος αριθμός βημάτων ανά διαδρομή|1837.63|12.98
| Μέσος αριθμός ανταμοιβών ανά κίνηση|-3.98|0.62

</center> 
